app/controllers/Promotion.scala

```diff
@@ -20,68 +20,81 @@ import views.support.Pricing._
+    tpBackend.promoService.findPromotionFuture(promoCode).flatMap { promotion =>
       promotion.filterNot(_.isTracking).fold {
-        NotFound(Json.obj("errorMessage" -> s"Sorry, we can't find that code."))
+        Future.successful(NotFound(Json.obj("errorMessage" -> s"Sorry, we can't find that code.")))
       } {
         promo =>
-          val result = promo.validateFor(prpId, country)
-          def body = Json.obj(
-            "promotion" -> Json.toJson(promo),
-            "adjustedRatePlans" -> Json.toJson(getAdjustedRatePlans(promo, country, currency)),
-            "isValid" -> result.isRight,
-            "errorMessage" -> result.swap.toOption.map(_.msg)
-          )
-          result.fold(_ => NotAcceptable(body), _ => Ok(body))
+          getAdjustedRatePlans(promo, country, currency).map { ratePlans =>
+            val result = promo.validateFor(prpId, country)
+
+            def body = Json.obj(
+              "promotion" -> Json.toJson(promo),
+              "adjustedRatePlans" -> Json.toJson(ratePlans),
+              "isValid" -> result.isRight,
+              "errorMessage" -> result.swap.toOption.map(_.msg)
+            )
+
+            result.fold(_ => NotAcceptable(body), _ => Ok(body))
+          }
       }
     }
   }
 
 
   def validate(promoCode: PromoCode, country: Country, currency: Option[Currency]) = NoCacheAction.async { implicit request =>
-    implicit val resolution: TouchpointBackend.Resolution = TouchpointBackend.forRequest(PreSigninTestCookie, request.cookies)
+    implicit val resolution: TouchpointBackends.Resolution = fBackendFactory.forRequest(PreSigninTestCookie, request.cookies)
     implicit val tpBackend = resolution.backend
 
-    tpBackend.promoService.findPromotionFuture(promoCode).map { promotion =>
+    tpBackend.promoService.findPromotionFuture(promoCode).flatMap { promotion =>
       promotion.fold {
-        NotFound(Json.obj("errorMessage" -> s"Sorry, we can't find that code."))
+        Future.successful(NotFound(Json.obj("errorMessage" -> s"Sorry, we can't find that code.")))
       } { promo =>
-        val result = promo.validate(country)
-        def body = Json.obj(
-          "promotion" -> Json.toJson(promo),
-          "adjustedRatePlans" -> Json.toJson(getAdjustedRatePlans(promo, country, currency)),
-          "isValid" -> result.isRight,
-          "errorMessage" -> result.swap.toOption.map(_.msg)
-        )
-        result.fold(_ => NotAcceptable/*!?! fix this*/(body), _ => Ok(body))
-      }
+        getAdjustedRatePlans(promo, country, currency).map { ratePlans =>
+          val result = promo.validate(country)
 
+          def body = Json.obj(
+            "promotion" -> Json.toJson(promo),
+            "adjustedRatePlans" -> Json.toJson(ratePlans),
+            "isValid" -> result.isRight,
+            "errorMessage" -> result.swap.toOption.map(_.msg)
+          )
+
+          result.fold(_ => NotAcceptable /*!?! fix this*/ (body), _ => Ok(body))
```

Comment:
github claims I added this (and I probably did) but not in this PR

Reply:
😆